File: haskell/prime-factors/src/PrimeFactors.hs
module PrimeFactors
  ( primeFactors
  ) where

import           Control.Monad (join)

primeFactors :: Integer -> [Integer]
primeFactors = factors prime
  where
    factors [] _       = undefined
    factors _ 1        = []
    factors pps@(p:ps) m
      | m `mod` p == 0 = p:factors pps (m `div` p)
      | otherwise      = factors ps m

prime :: [Integer]
prime = 2:3:[ x | k <- [1 ..]
                , r <- [-1, 1]
                , let x = 6 * k + r
                , all ((0 /=) . (x `mod`)) . takeWhile ((<= x) . join (*)) $ prime
                ]

File: cpp/prime-factors/prime_factors.cpp
#include "prime_factors.h"

std::vector<int> prime_factors::of(int n) {
    int divisor = 2;
    std::vector<int> v;

    while (n > 1) {
        if (n % divisor == 0) {
            v.push_back(divisor);
            n /= divisor;
        } else {
            divisor += divisor > 2 ? 2 : 1;
        }
    }

    return v;
}
